File: haskell/src/Day06.hs
module Day06 where

import Data.Set (Set, fromList, union, intersection)

type Answer = Char
type Group = [Set Answer]

day06 :: IO ()
day06 = do
    input <- readFile "../data/day06.txt"
    let groups = parseGroups input
    let answer1 = sum . map (length . answersByAnyone) $ groups
    let answer2 = sum . map (length . answersByEveryone) $ groups
    putStrLn "\n===== Day 6 ====="
    putStrLn $ "The sum of answers that anyone in the group affirmed is " ++ show answer1
    putStrLn $ "The sum of answers that everyone in the group affirmed is " ++ show answer2

parseGroups :: String -> [Group]
parseGroups = map (map fromList) . parseGroups' [] . lines
    where
        parseGroups' people [] = [people]
        parseGroups' people ("" : rest) = people : parseGroups' [] rest
        parseGroups' people (p : rest) = parseGroups' (p: people) rest

answersByAnyone :: Group -> Set Answer
answersByAnyone = foldl1 union

answersByEveryone :: Group -> Set Answer
answersByEveryone = foldl1 intersection

File: typescript/src/day06.ts

type Person = string;
type Group = Person[];

export default function day06(data : string) {
  const groups: Group[] = data.split('\n\n').map(g => g.split('\n'));
  const answersFromAnyone = groups.map(someoneAnswered);
  const answersFromEveryone = groups.map(everyoneAnswered);

  console.log("===== Day 6 =====");
  console.log("The sum of each groups answers is", sumCount(answersFromAnyone));
  console.log("The sum of answers from everyone in the group is", sumCount(answersFromEveryone));
}

function someoneAnswered(group: Group): string[] {
  const result = new Set<string>();
  group.forEach(p => p.split('').forEach(a => result.add(a)));
  return [...result];
}

function everyoneAnswered(group: Group) : string[]{
  const possibleAnswers = [...someoneAnswered(group)];
  const allAnswered = possibleAnswers.filter( answer => group.every(person => person.includes(answer)));
  return allAnswered;
}

function sumCount(as: { length: number }[]): number {
  return as.reduce((sum, a) => sum + a.length, 0);
}